File: javascript/hexcompress.js
"use strict";

function decompress(buffer) {
  var result = new Buffer(buffer.length * 2);
  var r = 0;

  for (var i=0; i < buffer.length; i++) {
    var byte = buffer[i];
    // 33: !, 126: ~
    if (byte !== 0 && (byte < 33 || byte > 126)) {
      var hex = byte.toString(16);
      if (byte < 16) {
        hex = "0" + hex;
      }
      result[r] = hex.charCodeAt(0);
      result[r + 1] = hex.charCodeAt(1);
      r += 2;
    }
    else {
      result[r] = byte;
      r++;
    }
  }

  return result.toString("ascii", 0, r);
}

module.exports = {
  decompress: decompress
};

File: clojure/src/io/torsten/hexcompress.clj
(ns io.torsten.hexcompress)

(def hex-bytes
  (set (map #(->> % (format "%x") .getBytes first byte) (range 16))))

(def printable-ascii
  (conj (set (map byte (range (int \!) 127))) (byte 0)))

(defn hex-encode
  "Takes to bytes and hex-encodes them into one
   ['f', 'f'] => 0xff."
  [a b]
  (Integer/parseInt (String. (byte-array [a b])) 16))

(defn drop-last-vec
  "Like drop-last but efficiently for vectors."
  [v]
  (let [c (count v)]
    (if (= c 0)
      v
      (subvec v 0 (- c 1)))))

(defn compress
  "Compress a string into bytes."
  [string]
  (byte-array
    (first
      (reduce
        (fn [[result prev] next]
          (if (and (hex-bytes prev)
                   (hex-bytes next)
                   (not (printable-ascii (hex-encode prev next))))
            [(conj (drop-last-vec result) (hex-encode prev next)) nil]
            [(conj result next) next]))
        [[] nil]
        (.getBytes string)))))

(defn decompress
  "Decompress bytes into a new string."
  [bytes]
  (new String
    (byte-array
      (mapcat
        (fn [byte]
          (if (not (printable-ascii byte))
            (map int (format "%02x" byte))
            [byte]))
        bytes))))
